### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by KissSheepa** [WuJH2001/swift4d#6]
Dear author, first of all, thank you very much for open-sourcing this influential work. However, when I reproduced the project code, I found that the performance tested under the N3DV dataset, especially the PSNR, was 2 to 3 points lower in each scene, which is very different from the evaluation data in the checkpoints file you provided. What is the reason? Do the relevant parameter configurations need to be modified?The picture shows my own test results of cook_spinach and the results in checkpoints

![Image](https://github.com/user-attachments/assets/b1aeb43e-b459-490f-9a2d-0154b43dabaf)

![Image](https://github.com/user-attachments/assets/488d2e19-2a95-44cf-9e5b-1c6d0c531be8)

**Comment by WuJH2001** [comment]
> Hello, I’ve encountered the same issue. The performance is similar to the above, with the following results on cook_spinach: Avg PSNR: 31.3613338 Avg DSSIM1: 0.0292308 Avg DSSIM2: 0.0000040 Avg LPIPS: 0.0440804 I’m new to the field of dynamic scene reconstruction, so I’m still trying to understand the details. As I understand it, issues such as floating objects occluding the test camera or "foggy" views in certain viewpoints of the N3DV dataset are inherent problems with the dataset (which should also affect other methods). Therefore, these factors do not seem to be directly responsible for the difficulty in reproducing the results reported in the paper. I want to know how to reproduce the results given in the paper. Looking forward to your reply, thank you. Looking forward to your response. Thank you!


---

Hello, thank you for your interest in our project. We’re very sorry for the issue you encountered. To investigate further, we retested our code several times as shown in the figure below (each test involved 300 frames). According to the results, we did not encounter the problem you described (i.e., such a large PSNR gap).

Regarding the fog observed from certain viewpoints in the **cook\_spinach** scene, we recommend modifying line 329 in `train.py` by removing the condition: `and stage == "coarse"`. This allows the model to periodically prune floaters from test views, leading to more stable PSNR values and avoiding the issue you experienced.

We also found that another possible reason for the large PSNR difference could be a color bias caused by the fog in some views. This might cause the test views to lean toward the foggy color tones, resulting in a significant PSNR drop.

If possible, could you share the rendered images from the test viewpoints? This would help us provide more targeted support.

![Image](https://github.com/user-attachments/assets/f58e289c-cf5d-4de4-9892-d4cc794b4fdd)

![Image](https://github.com/user-attachments/assets/4903d755-bfd6-4f46-825e-4c44f039fb92)
---

**Comment by CBY-9527** [comment]
> > Hello, I’ve encountered the same issue. The performance is similar to the above, with the following results on cook_spinach: Avg PSNR: 31.3613338 Avg DSSIM1: 0.0292308 Avg DSSIM2: 0.0000040 Avg LPIPS: 0.0440804 I’m new to the field of dynamic scene reconstruction, so I’m still trying to understand the details. As I understand it, issues such as floating objects occluding the test camera or "foggy" views in certain viewpoints of the N3DV dataset are inherent problems with the dataset (which should also affect other methods). Therefore, these factors do not seem to be directly responsible for the difficulty in reproducing the results reported in the paper. I want to know how to reproduce the results given in the paper. Looking forward to your reply, thank you. Looking forward to your response. Thank you!
> 
> Hello, thank you for your interest in our project. We’re very sorry for the issue you encountered. To investigate further, we retested our code several times as shown in the figure below (each test involved 300 frames). According to the results, we did not encounter the problem you described (i.e., such a large PSNR gap).
> 
> Regarding the fog observed from certain viewpoints in the **cook_spinach** scene, we recommend modifying line 329 in `train.py` by removing the condition: `and stage == "coarse"`. This allows the model to periodically prune floaters from test views, leading to more stable PSNR values and avoiding the issue you experienced.
> 
> We also found that another possible reason for the large PSNR difference could be a color bias caused by the fog in some views. This might cause the test views to lean toward the foggy color tones, resulting in a significant PSNR drop.
> 
> If possible, could you share the rendered images from the test viewpoints? This would help us provide more targeted support.
> 
> ![Image](https://github.com/user-attachments/assets/e9c5e51d-359a-4dff-8311-7b4108a64d5d)
> 
> ## ![Image](https://github.com/user-attachments/assets/4903d755-bfd6-4f46-825e-4c44f039fb92)

非常感谢您的回复，以下是我跑出的结果：

![Image](https://github.com/user-attachments/assets/20f9dfc4-aab4-41dc-bc98-28c6aabe20e9)
![Image](https://github.com/user-attachments/assets/2da23a2c-1ff1-475d-a9e8-65d35e892d2a)

我是按照说明文档中的步骤：生成点云-->降采样-->训练。我发现我的打印结果中的点数量要明显比您给出的要少很多（点的个数低于1000，而您给出的图片中点的数量有38780），我猜想性能差异的主要原因应该就是生成点的数量导致的。不知是否能解惑一下为什么我生成的点会这么少（我没有做任何参数的修改），以及如何修改代码可以让第一步生成的点数增加。   期待回复，谢谢！

**Comment by WuJH2001** [comment]
处理数据这一步我们也是跟随[4DGS](https://github.com/hustvl/4DGaussians) 的处理流程，按道理来说对于这种场景生成的稠密点云文件 fused.py 里面的点云数量应该大于10w，您使用colmap跑出来的几百个点云数量确实存在问题。您可以检查一下图像数据是否处理正确，有没有存在某些相机视角未被处理。如果确认无误可以尝试再次运行代码（可以试试[4DGS](https://github.com/hustvl/4DGaussians) 代码库里面的处理流程），因为colmap本身是一个不确定的过程，每次运行都会出现不一样的结果，可能某次运行会导致视角无法匹配，也可能全部匹配上，期待回复！

**Comment by CBY-9527** [comment]
> 处理数据这一步我们也是跟随[4DGS](https://github.com/hustvl/4DGaussians) 的处理流程，按道理来说对于这种场景生成的稠密点云文件 fused.py 里面的点云数量应该大于10w，您使用colmap跑出来的几百个点云数量确实存在问题。您可以检查一下图像数据是否处理正确，有没有存在某些相机视角未被处理。如果确认无误可以尝试再次运行代码（可以试试[4DGS](https://github.com/hustvl/4DGaussians) 代码库里面的处理流程），因为colmap本身是一个不确定的过程，每次运行都会出现不一样的结果，可能某次运行会导致视角无法匹配，也可能全部匹配上，期待回复！

我遇到了一个小问题，在执行python scripts/preprocess_dynerf.py --datadir data/dynerf/cook_spinach这一步的时候，neural_3D_dataset_NDC.py中的  “frame = np.stack(frames, axis=0)”一行会报错，提示frames是空的。
preprocess_dynerf.py中先执行 train_dataset = Neural3D_NDC_Dataset(....)，训练的视频是从1开始的，而neural_3D_dataset_NDC.py中的self.calc_std是从视频0开始计算，此时还未对视频0进行抽帧。   我尝试修改代码以正常运行，但是发现最终的性能不佳，想问您一下应该如何正确修正该问题。期待回复，谢谢！

**Comment by WuJH2001** [comment]
非常抱歉，由于我们一开始使用 4DGS 代码处理好了数据之后，修改了他的部分数据代码，但是忘记了修复这个不一致的 bug。如果您比较紧急的话，我们推荐您可以直接使用 4DGS 的数据处理代码，再运行我们的训练代码是一样的效果。感谢反馈，我们将更新我们的代码。

**Comment by CBY-9527** [comment]
> 非常抱歉，由于我们一开始使用 4DGS 代码处理好了数据之后，修改了他的部分数据代码，但是忘记了修复这个不一致的 bug。如果您比较紧急的话，我们推荐您可以直接使用 4DGS 的数据处理代码，再运行我们的训练代码是一样的效果。感谢反馈，我们将更新我们的代码。

非常感谢您的回复。我重新下载了3.10版本的colmap（之前安装的是最新版本3.11.1），并编译安装，生成的点数量再降采样后和您图片中的差不多。以下是我重跑的spinach的结果，想咨询一下：1. 在约6000次迭代时性能最佳，不知该结果是否正常；2. 最终结果取的是13000迭代时的结果还是取性能最高时的结果呢。期待回复，谢谢！

![Image](https://github.com/user-attachments/assets/b3dbe442-8e5e-44cd-94ad-f1a1c74ca5ea)

**Comment by WuJH2001** [comment]
> 非常感谢您的回复。我重新下载了3.10版本的colmap（之前安装的是最新版本3.11.1），并编译安装，生成的点数量再降采样后和您图片中的差不多。以下是我重跑的spinach的结果，想咨询一下：1. 在约6000次迭代时性能最佳，不知该结果是否正常；2. 最终结果取的是13000迭代时的结果还是取性能最高时的结果呢。期待回复，谢谢！

由于仅需建模少量动态区域，我们的方法可以实现更快收敛。我们实验过程也发现，在约7000轮次时性能已趋于稳定，因此在6000轮次达到最佳性能是正常的。为保持与先前方法一致，我们设定训练轮次为13000，但在实践中通常采用7000和13000中表现更优的轮数，以避免过拟合到训练视角导致的客观质量下降。
